File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        lst = []
        open_bracket = "({["

        for char in s:
            if char in open_bracket:
                lst.append(char)
            else:
                if not lst:
                    return False
                if char == ")" and lst.pop() != "(":
                    return False
                elif char == "]" and lst.pop() != "[":
                    return False
                elif char == "}" and lst.pop() != "{":
                    return False

        return len(lst) == 0

File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        unordered_map<char, char> closeToOpen;
        closeToOpen[')'] = '(';
        closeToOpen['}'] = '{';
        closeToOpen[']'] = '[';
        for(char ch: s) {
            if(ch == '(' || ch == '{' || ch == '[') {
                stk.push(ch);
            } else {
                if(stk.empty()) return false;

                char close = ch;
                char open = closeToOpen[ch];

                // Close should match to corresponding open
                if(open == stk.top()) {
                    stk.pop();
                } else {
                    return false;
                }
            }
        }
        return stk.empty();
    }
};